File: lib/encrypt.ts
import crypto from 'crypto';

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 12;
const AUTH_TAG_LENGTH = 16;

function getKey(): Buffer {
  const secretKey = process.env.SECRET_KEY;
  if (!secretKey) throw new Error('SECRET_KEY is missing in environment variables');
  if (secretKey.length !== 32) throw new Error('SECRET_KEY must be 32 bytes');
  return Buffer.from(secretKey, 'utf8');
}

const encrypt = (text: string): string => {
  const key = getKey();
  const iv = crypto.randomBytes(IV_LENGTH);
  const cipher = crypto.createCipheriv(ALGORITHM, key, iv, { authTagLength: AUTH_TAG_LENGTH });

  let encrypted = cipher.update(text, 'utf8', 'hex');
  encrypted += cipher.final('hex');
  const authTag = cipher.getAuthTag().toString('hex');

  // Format: iv:authTag:ciphertext (all hex)
  return `${iv.toString('hex')}:${authTag}:${encrypted}`;
};

export { encrypt };

File: python_backend/encrypt.py
import os
from Crypto.Cipher import AES
from dotenv import load_dotenv

load_dotenv()

secret_key = os.getenv('SECRET_KEY')

if not secret_key:
    raise ValueError('SECRET_KEY is missing in environment variables')

if len(secret_key) != 32:
    raise ValueError('SECRET_KEY must be 32 bytes')

key = bytes(secret_key, 'utf-8')

def encrypt(text: str) -> str:
    """Encrypt using AES-256-GCM with random IV. Format: iv:authTag:ciphertext (all hex)"""
    iv = os.urandom(12)
    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    ciphertext, auth_tag = cipher.encrypt_and_digest(text.encode('utf-8'))
    return f"{iv.hex()}:{auth_tag.hex()}:{ciphertext.hex()}"